rust/noosphere-ns/src/name_system.rs

```diff
@@ -0,0 +1,195 @@
+    /// but can be manually called push updated records to the network.
+    ///
+    /// Can fail if NameSystem is not connected or if no peers can be found.
+    pub async fn propagate_records(&self) -> Result<()> {
+        if self.dht.is_none() {
+            return Err(anyhow!("not connected"));
+        }
+        if self.hosted_records.is_empty() {
+            return Ok(());
+        }
+
+        let pending_tasks: Vec<_> = self
+            .hosted_records
+            .iter()
+            .map(|(identity, record)| self.dht_set_record(identity, record))
+            .collect();
+        try_join_all(pending_tasks).await?;
+        Ok(())
+    }
+
+    /// Propagates the corresponding managed sphere's content Cid on nearby peers
+    /// in the DHT network.
+    ///
+    /// Can fail if NameSystem is not connected or if no peers can be found.
+    pub async fn set_record(&mut self, identity: &String, link: &Cid) -> Result<()> {
+        if self.dht.is_none() {
+            return Err(anyhow!("not connected"));
+        }
+
+        let record = NSRecord::new_from_ttl(link.to_owned(), self.ttl)?;
+        self.dht_set_record(identity, &record).await?;
+        self.hosted_records.insert(identity.to_owned(), record);
+        Ok(())
+    }
+
+    /// Gets the content Cid for the provided sphere identity.
+    pub async fn get_record(&mut self, identity: &String) -> Option<&Cid> {
+        // Round about way of checking for local valid record before
+        // querying the DHT network due to the borrow checker.
+        // https://stackoverflow.com/questions/70779967/rust-borrow-checker-and-early-returns#
+        let has_valid_record: bool = {
+            if let Some(record) = self.resolved_records.get(identity) {
+                !record.is_expired()
+            } else {
+                false
+            }
+        };
+
+        // No non-expired record found locally, query the network.
+        if !has_valid_record {
+            match self.dht_get_record(identity).await {
+                Ok((_, Some(record))) => {
+                    self.resolved_records.insert(identity.to_owned(), record);
+                }
+                Ok((_, None)) => {}
+                Err(_) => {}
+            }
+        }
+        match self.resolved_records.get(identity) {
+            Some(record) => Some(&record.cid),
+            None => None,
+        }
+    }
+
+    /// Clears out the internal cache of resolved records.
+    pub fn flush_records(&mut self) {
+        self.resolved_records.drain();
+    }
+
+    /// Clears out the internal cache of resolved records
+    /// for the matched identity. Returned value indicates whether
+    /// a record was successfully removed.
+    pub fn flush_records_for_identity(&mut self, identity: &String) -> bool {
+        self.resolved_records.remove(identity).is_some()
+    }
+
+    pub fn p2p_address(&self) -> Option<&Multiaddr> {
```

Comment:
Unused currently; previously used to log listening address in CLI, which seems useful, so leaving it here for now